From 1cf2a3b871db8d9846bbc7334aa84d32b99a837a Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Mon, 22 Sep 2025 15:28:53 +0200
Subject: [PATCH] Issue 7012 - improve dscrl dbverify result when backend does
 not exists (#7013)

* Issue 7012 - improve dscrl dbverify result when backend does not exists

Improve error message when running dsctl instance dbverify wrongBeName
To tell explicitly that the backend does not exists in bdb case
To tell explicitly that dbverify is useless and always successfull in lmdb case

Issue: 7012

Reviewed by: @mreynolds389 and @tbordaz (Thanks!)
---
 dirsrvtests/tests/suites/clu/dbverify_test.py | 44 +++++++++++++++++++
 .../slapd/back-ldbm/db-bdb/bdb_verify.c       |  2 +
 .../slapd/back-ldbm/db-mdb/mdb_verify.c       |  1 +
 3 files changed, 47 insertions(+)

diff --git a/dirsrvtests/tests/suites/clu/dbverify_test.py b/dirsrvtests/tests/suites/clu/dbverify_test.py
index fe9b1a6ed..028afdbdb 100644
--- a/dirsrvtests/tests/suites/clu/dbverify_test.py
+++ b/dirsrvtests/tests/suites/clu/dbverify_test.py
@@ -31,6 +31,7 @@ def set_log_file(request):
 
     def fin():
         log.info('Delete log file')
+        log.removeHandler(fh)
         os.remove(LOG_FILE)
 
     request.addfinalizer(fin)
@@ -68,6 +69,49 @@ def test_dsctl_dbverify(topology_st, set_log_file):
         assert message in file_content
 
 
+def run_dbverify(inst, bename, rc=0):
+    # Run dsctl inst dbverify bename and check it returns rc
+    cmd = [ 'dsctl', inst.serverid, 'dbverify', bename ]
+    log.info(f'Run: {cmd}')
+    result =  subprocess.run(cmd, encoding='utf-8', capture_output=True, text=True)
+    log.info(f'Returned: {result.returncode} stdout: {result.stdout} stderr: {result.stderr}')
+    assert result.returncode == rc
+    return result
+
+
+
+def test_dbverify_bad_bename(topology_st, set_log_file):
+    """Test dbverify tool when backend does not exists
+
+    :id: da79a4cc-956c-11f0-8220-c85309d5c3e3
+    :setup: Standalone instance
+    :steps:
+         1. Create DS instance
+         2. Run dbverify
+         3. Check if dbverify provides the expected messages
+    :expectedresults:
+         1. Success
+         2. Success
+         3. Success
+    """
+
+    standalone = topology_st.standalone
+    if get_default_db_lib() == 'bdb':
+        message = "dbverify failed"
+        logmessage = "Backend 'userFoot' does not exist."
+        rc = 1
+    else:
+        message = "dbverify successful"
+        logmessage = "db_verify feature is meaningless"
+        rc = 0
+
+    standalone.stop()
+    result = run_dbverify(standalone, 'userFoot', rc)
+    assert message in result.stdout
+    assert logmessage in result.stderr
+
+
+
 if __name__ == '__main__':
     # Run isolated
     # -s for DEBUG mode
diff --git a/ldap/servers/slapd/back-ldbm/db-bdb/bdb_verify.c b/ldap/servers/slapd/back-ldbm/db-bdb/bdb_verify.c
index 2859dad9d..3af6bb2c6 100644
--- a/ldap/servers/slapd/back-ldbm/db-bdb/bdb_verify.c
+++ b/ldap/servers/slapd/back-ldbm/db-bdb/bdb_verify.c
@@ -195,6 +195,8 @@ bdb_verify(Slapi_PBlock *pb)
                 }
                 rval_main |= bdb_dbverify_ext(inst, verbose);
             } else {
+                slapi_log_err(SLAPI_LOG_ERR, "bdb_verify",
+                              "Backend '%s' does not exist.\n", *inp);
                 rval_main |= 1; /* no such instance */
             }
         }
diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_verify.c b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_verify.c
index 4d5df1bfc..43bacba9f 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_verify.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_verify.c
@@ -17,5 +17,6 @@
 int
 dbmdb_verify(Slapi_PBlock *pb)
 {
+    slapi_log_err(SLAPI_LOG_WARNING, "dbmdb_verify", "With lmdb, db_verify feature is meaningless and is always successfull.\n");
     return 0;  /* Fonction useless with lmdb - as we can verify the db when doing a backup */
 }
-- 
2.52.0

